### DISCUSSION TITLE
CMake + PCH + ccache with touched headers

### DISCUSSION THREAD

**Discussion by m-dot-jung** [ccache/ccache#1218] [Support]
Hi,

we are trying to use CCache in an gitlab runner environment. 
This means that for each new build, all mtimes will be newer than for any build before as a clean git checkout happens, which sets all mtimes. Is there a way to disable checks of mtime?

We cannot seem to get the precompiled header to not recompile - always it states

Looking for object file hash in /tmp/ccache/2/3/c2fe9a28bbc28f35fe8a41af31cfba-831993.manifest
**Precompiled header includes ../source/project/include/TouchedHeader.h, which has a new mtime**
Did not find object file hash in manifest
Running real compiler
Executing /usr/bin/c++ ...
Result: cache miss

As the precompiled header is rebuilt (which is included everywhere) a full rebuild is triggered.

If I try to recreate the same behavior, I run cmake and build, remove the build folder, run again - everything works nicely with ccache. Simply running `touch -m TouchedHeader.h` on a file that is included in the precompiled header will rebuild everything. Is there a way to check these for content changes, as well? We can guarantee that there is no racing situation...

Command line:
```
ccache /usr/bin/c++ 
-D<Some definitions> 
-I<Some include directories> 
-W<some warnings> 
-fpch-preprocess -O2 -g -DNDEBUG -std=c++17 -Winvalid-pch 
-include /builds/project/build/core-lib/CMakeFiles/core-lib.dir/cmake_pch.hxx 
-MD 
-MT core-lib/CMakeFiles/core-lib.dir/source/Test.cpp.o 
-MF core-lib/CMakeFiles/core-lib.dir/source/Test.cpp.o.d 
-o core-lib/CMakeFiles/core-lib.dir/source/Test.cpp.o 
-c /builds/project/source/core-lib/source/Test.cpp
```

This is our CCache Configuration: [ccache.conf](https://github.com/ccache/ccache/files/9991480/ccache.conf.log)

**Comment by jrosdahl** [comment]
> We cannot seem to get the precompiled header to not recompile - always it states
> 
> Looking for object file hash in /tmp/ccache/2/3/c2fe9a28bbc28f35fe8a41af31cfba-831993.manifest 
> **Precompiled header includes ../source/project/include/TouchedHeader.h, which has a new mtime**
> Did not find object file hash in manifest
> Running real compiler Executing /usr/bin/c++ ...
> Result: cache miss

Clang stores the mtimes of the included files in the PCH and will error out if a header with a different mtime is later used without rebuilding the PCH. This is why ccache will produce a cache miss instead of producing a PCH that Clang will choke on.

However, this was improved in #624 (included in ccache 4.0): you can compile with `-fno-pch-timestamp` to tell Clang not to include mtimes, and ccache should then work. Unfortunately, that PR did not update the documentation to mention this. I've done that now in 682f58a0e1d45d691970944117d893bdb523fc3e.

**Reply by m-dot-jung** [reply]
Hey, thanks for the answer, we are compiling with gcc not with Clang, `-fno-pch-timestamp`, which gives:
c++: error: unrecognized command-line option `-fno-pch-timestamp`

**Reply by jrosdahl** [reply]
Ah. Then the problem is that ccache is unable to detect that `/usr/bin/c++` is GCC. This can be fixed by setting [`compiler_type = gcc`](https://ccache.dev/manual/4.7.3.html#config_compiler_type).

**Reply by m-dot-jung** [reply]
Specifying the `compiler_type`-parameter didn't help as we are running on ccache version 3.4.2, perhaps we should upgrade. However, not using the symbolic links, but the executables directly did the trick (pointing directly via `-DCMAKE_CXX_COMPILER=g++-10 -DCMAKE_C_COMPILER=gcc-10` instead of the symbolic link `/usr/bin/c++`).

**Reply by jrosdahl** [reply]
> Specifying the compiler_type-parameter didn't help as we are running on ccache version 3.4.2, perhaps we should upgrade. However, not using the symbolic links, but the executables directly did the trick [...]

With ccache 4.1+ it will work out of the box to use `/usr/bin/c++` as a symlink to the compiler:
https://github.com/ccache/ccache/blob/69446c4d3d9ff793696fa3f0118638198c5efb56/doc/NEWS.adoc#ccache-41

**Comment by afbjorklund** [comment]
There are some scripts that will reset the mtime to match the timestamps in git, this is useful for getting cache hits (where valid):

https://stackoverflow.com/questions/1964470/whats-the-equivalent-of-subversions-use-commit-times-for-git/13284229#13284229

https://github.com/MestreLion/git-tools/blob/main/git-restore-mtime

**Reply by m-dot-jung** [reply]
Thanks for pointing out the scripts. However, I was under the impression that using this work-around, but not comparing the file contents, would lead to situations where switching back to an older branch would not trigger correct recompilation.

**Reply by afbjorklund** [reply]
That is true, it was assuming a clean build directory but a shared cache.

**Comment by silverqx** [comment]
have asked the same question on the CMake forum - [Ccache clang and -fno-pch-timestamp](https://discourse.cmake.org/t/ccache-clang-and-fno-pch-timestamp/7253).

The question is how can we add the `-fno-pch-timestamp` compiler option only for the precompiled header compiler command?
